File: shared/ios/ReleafCore/Sources/ReleafCoreScan/BusinessCard/ScoringWeights.swift
/*
 * ScoringWeights.swift
 *
 * Centralised tunables for the business-card extraction pipeline.
 * Mirror of `ScoringWeights.kt` — same defaults, same semantics, so
 * benchmark grids run identically across both platforms.
 */

import Foundation

public struct ScoringWeights: Sendable {
    public let emailBase: Double
    public let phoneBase: Double
    public let websiteBase: Double
    public let designationBase: Double
    public let nameBase: Double
    public let companyBase: Double
    public let addressBase: Double
    public let postcodeBase: Double

    public let largeTextBonusMax: Double
    public let topPositionBonusMax: Double
    public let nameDesignationAdjacencyBonus: Double
    public let adjacencyYDistance: Double

    public let designationVocabBonus: Double
    public let companySuffixBonus: Double

    public let engineConfidenceWeight: Double

    public let nameDigitsPenalty: Double
    public let nameTokenCountPenalty: Double
    public let companyPunctuationPenalty: Double
    public let duplicateOverlapPenalty: Double
    /// Penalty applied to NAME candidates whose block height is at
    /// or near `layout.maxHeight`. Mirror of the Android weight.
    public let largestTextPenaltyForName: Double
    /// Bonus when the block's first token is a known honorific
    /// ("Mr", "Mrs", "Dr", "Sri", "Smt", …). Mirror of the Android
    /// weight.
    public let nameSalutationBonus: Double

    public let minNameScore: Double
    public let minCompanyScore: Double
    public let minDesignationScore: Double
    public let minAddressScore: Double
    public let minPhoneScore: Double
    public let minEmailScore: Double
    public let minWebsiteScore: Double

    public init(
        emailBase: Double                      = 10.0,
        phoneBase: Double                      = 9.0,
        websiteBase: Double                    = 8.0,
        designationBase: Double                = 5.0,
        nameBase: Double                       = 4.0,
        companyBase: Double                    = 4.0,
        addressBase: Double                    = 4.0,
        postcodeBase: Double                   = 6.0,
        largeTextBonusMax: Double              = 6.0,
        topPositionBonusMax: Double            = 6.0,
        nameDesignationAdjacencyBonus: Double  = 4.0,
        adjacencyYDistance: Double             = 0.10,
        designationVocabBonus: Double          = 5.0,
        companySuffixBonus: Double             = 5.0,
        engineConfidenceWeight: Double         = 4.0,
        nameDigitsPenalty: Double              = 10.0,
        nameTokenCountPenalty: Double          = 5.0,
        companyPunctuationPenalty: Double      = 3.0,
        duplicateOverlapPenalty: Double        = 2.0,
        largestTextPenaltyForName: Double      = 4.0,
        nameSalutationBonus: Double            = 6.0,
        minNameScore: Double                   = 5.0,
        minCompanyScore: Double                = 5.0,
        minDesignationScore: Double            = 5.0,
        minAddressScore: Double                = 4.0,
        minPhoneScore: Double                  = 6.0,
        minEmailScore: Double                  = 8.0,
        minWebsiteScore: Double                = 6.0
    ) {
        self.emailBase                         = emailBase
        self.phoneBase                         = phoneBase
        self.websiteBase                       = websiteBase
        self.designationBase                   = designationBase
        self.nameBase                          = nameBase
        self.companyBase                       = companyBase
        self.addressBase                       = addressBase
        self.postcodeBase                      = postcodeBase
        self.largeTextBonusMax                 = largeTextBonusMax
        self.topPositionBonusMax               = topPositionBonusMax
        self.nameDesignationAdjacencyBonus     = nameDesignationAdjacencyBonus
        self.adjacencyYDistance                = adjacencyYDistance
        self.designationVocabBonus             = designationVocabBonus
        self.companySuffixBonus                = companySuffixBonus
        self.engineConfidenceWeight            = engineConfidenceWeight
        self.nameDigitsPenalty                 = nameDigitsPenalty
        self.nameTokenCountPenalty             = nameTokenCountPenalty
        self.companyPunctuationPenalty         = companyPunctuationPenalty
        self.duplicateOverlapPenalty           = duplicateOverlapPenalty
        self.largestTextPenaltyForName         = largestTextPenaltyForName
        self.nameSalutationBonus               = nameSalutationBonus
        self.minNameScore                      = minNameScore
        self.minCompanyScore                   = minCompanyScore
        self.minDesignationScore               = minDesignationScore
        self.minAddressScore                   = minAddressScore
        self.minPhoneScore                     = minPhoneScore
        self.minEmailScore                     = minEmailScore
        self.minWebsiteScore                   = minWebsiteScore
    }
}

File: shared/android/shared/scan/src/main/kotlin/app/releaf/shared/scan/businesscard/ScoringWeights.kt
/*
 * ScoringWeights.kt
 *
 * Centralised tunables for the business-card extraction pipeline.
 * Every magic number lives here — classifiers use the typed accessors,
 * and the benchmark harness can drive grid searches over alternative
 * weight sets without reaching into individual classifiers.
 *
 * Defaults are calibrated for typical Indian business cards (English-
 * Latin, mixed-script names, +91 / 10-digit phone numbers). Callers
 * with different priors can supply their own instance to
 * [BusinessCardExtractor.extract].
 */

package app.releaf.shared.scan.businesscard

data class ScoringWeights(
    // Per-kind base scores. Higher base = "this kind has a stronger
    // independent signal". Email's regex hit is near-certain, name
    // requires multiple weak signals to add up.
    val emailBase: Double       = 10.0,
    val phoneBase: Double       = 9.0,
    val websiteBase: Double     = 8.0,
    val designationBase: Double = 5.0,
    val nameBase: Double        = 4.0,
    val companyBase: Double     = 4.0,
    val addressBase: Double     = 4.0,
    val postcodeBase: Double    = 6.0,

    // Layout bonuses. Top-of-card and large-font text are strong
    // priors for name / company. Caps at the listed values so a
    // single ultra-tall block doesn't dominate.
    val largeTextBonusMax: Double  = 6.0,
    val topPositionBonusMax: Double = 6.0,
    /**
     * Bonus applied when a designation candidate sits directly below
     * a name candidate (within `adjacencyYDistance` on the y-axis).
     * Captures the "Name / Title" stacking that's near-universal on
     * cards.
     */
    val nameDesignationAdjacencyBonus: Double = 4.0,
    val adjacencyYDistance: Double            = 0.10,

    // Vocabulary bonuses.
    val designationVocabBonus: Double = 5.0,
    val companySuffixBonus: Double    = 5.0,

    // Engine confidence multiplier — Apple Vision exposes per-block
    // confidence; ML Kit Play-Services 19.0.1 emits null. Treated as
    // neutral (0) when null so Android isn't penalised by absence
    // of signal.
    val engineConfidenceWeight: Double = 4.0,

    // Penalties — applied to specific (kind, signal) combinations.
    val nameDigitsPenalty: Double         = 10.0,
    val nameTokenCountPenalty: Double     = 5.0   // > 4 words doesn't read as a name
    ,
    val companyPunctuationPenalty: Double = 3.0,
    val duplicateOverlapPenalty: Double   = 2.0,

    // Selection thresholds. Candidates below the per-kind minimum
    // are dropped; pipeline emits null for the field (or skips the
    // entry for multi-value fields).
    val minNameScore: Double        = 5.0,
    val minCompanyScore: Double     = 5.0,
    val minDesignationScore: Double = 5.0,
    val minAddressScore: Double     = 4.0,
    val minPhoneScore: Double       = 6.0,
    val minEmailScore: Double       = 8.0,
    val minWebsiteScore: Double     = 6.0,
)
